### DISCUSSION TITLE
Random findings

### DISCUSSION THREAD

**Discussion by ghost** [city41/smaghetti#86] [General]
- Right Goomba pipe generator should only need you to be to the right of it to activate.
- The second byte of "screen y boundary" seems to control how many screens the room has vertically. I know you're working on level settings so maybe this was already figured out?
- Object Bank 50, ID 76 controls the background's y position. Changing the 5th byte definitely change the background's y position, though I'm not 100% sure how the number is handled.
- Second byte in unknown2 controls the y position of the "extra background" (clouds, water, etc.). So if you had a pool of water as your x effect, and unknown2 byte 2 was set to 10, the water pool would be at y=10 (16 blocks from the screen).
- 016C stops the parachute enemy generator and when you go to the left of it, it will spawn parachute enemies, and the same values (5-byte, 5th byte controls amount allowed onscreen, x position decides whether it spawns galoombas or bob-ombs).
- Supposedly 016F is also related to 016E, since they react to other generator sprites, but I still have no idea what they do. But I guess they're not specifically for the parachute enemies.
- 016A is the SMW Bullet Bill generator, used in "Wild Ride in the Sky," which activates when you go to the right of it. 0169 stops the Bullet Bill generator if you go to the right, and restarts it when you go to the left.
- If you place 0169 or 016A at an even x-position (x=0), the generator will shoot Bullet Bills in cardinal directions (which is an unused SMW leftover). if at an odd x-position (x=1), it will shoot the diagonal Bullet Bills. All Bullet Bills shot by it use graphicSets [-1, -1, -1, -1, 4, -1].
- 0171 is the Bowser fire spawner, not sure what sprite actually stops it. I don't think it actually has a generic stopper in the level it's used in.
- The only other 4-byte sprites in Classic 1-4 are the axe and 0129, which is a camera-related sprite that prevents you from going to the left of its x position once you pass it.
- 0075 is Bowser's fireball by itself, though some of the sprite lists also call it Lemmy's ball? Wonder how you change that, unless it's locked to the world it's in (even though the Koopalings aren't unlike in earlier SMB3s?)

**Comment by city41** [comment]
thank you! 

I just promoted goomba generator to finished. I'll eventually get all this stuff plus all the other findings everyone has been making into the editor, but I've got quite the  backlog now

**Reply by city41** [reply]
> The second byte of "screen y boundary" seems to control how many screens the room has vertically. I know you're working on level settings so maybe this was already figured out?

I think screen y boundary is a 16 byte value of the level's height in pixels. That does seem to work out at least for classic 1-1 and classic 1-2. Very strange they'd suddenly use pixels instead of tiles though, so I might be wrong.

**Comment by city41** [comment]
If anyone reading is interested, I am trying to figure out the level settings bytes related to level height, camera and mario's position. So far I've not made too much progress. Once these get figured out, smaghetti will enable putting mario where the user wants (limited by mario's position being defined in bytes, so 255 max for x and y), and sizing rooms

I've been dumping my findings here: https://github.com/city41/smaghetti/wiki/Level-Settings-(height,-x-y,-camera)

**Comment by city41** [comment]
For bullet bill generator `16a`, it is y being even or odd that determines diagonal vs cardinal, rather than x.

So far, I haven't been able to get `169` to do anything.

**Reply by nintyalex** [reply]
No, bumptytobumpty was right; it's the x value that determines how the Bullet Bills are shot. I just tested it.

**Reply by city41** [reply]
hmm, could it be both? I'm pretty sure y also influences their direction.

BulletBillGeneratorCardinal is available in the unfinished tab.

**Reply by city41** [reply]
yes, you're both right, x does influence direction. but I think y does too. one sec, let me try in hex-tree.

**Reply by city41** [reply]
I could have swore y influenced it, but I can't repro. Cool, looks like I was wrong. Glad I kept it in unfinished :)

anyone have any luck with bank 1, id 69? bumpty mentioned that's also a bullet bill generator, but I can't get it to do anything so far.

**Reply by nintyalex** [reply]
No, he said that sprite stops the Bills from firing.
BTW, bank 1, id 67 is what stops infinite Cheep Cheeps (both leaping and Spiny) from appearing.

**Reply by city41** [reply]
ah got it. I was confused by "If you place 0169 or 016A at an even x-position (x=0), the generator will shoot Bullet Bills in cardinal directions" and missed the sentence above that. Cool.

I haven't added any generator stoppers yet, still thinking a bit on how I want to do that. I was thinking of having "generator zones" where the user can pick a span to indicate start/stop.

**Comment by OQUENDOMAN** [comment]


**Reply by OQUENDOMAN** [reply]
a random finfing

**Reply by devtronic** [reply]
Please credit If you use research results of others 🙄 
This is just a translated Version of mine https://board.romhackersworld.eu/thread/13303-super-mario-advance-4-research-thread/

**Reply by city41** [reply]
Hey @OQUENDOMAN Taking others work as your own is not cool. Sorry @devtronic, he's not a member here. I will delete his post.

**Reply by OQUENDOMAN** [reply]
i don't never say was mine
im only posted because it so interesting your post @devtronic

**Reply by city41** [reply]
@devtronic I added your doc to the repo [here](https://github.com/city41/smaghetti/blob/main/reference/devtronicResearch.md). I translated it to English using Google.

Thanks for writing this! There looks to be lots of useful information!

**Comment by city41** [comment]
I believe `Second byte in unknown2 controls the y position of the "extra background" (clouds, water, etc.). So if you had a pool of water as your x effect, and unknown2 byte 2 was set to 10, the water pool would be at y=10 (16 blocks from the screen).` should allow adding the background that has the brown river of water that airships ride in. What's cool is you can swim in the water.

I think for starters, Smaghetti would just stick the river down towards the bottom of the room, but ultimately letting people decide would be ideal.

**Comment by devtronic** [comment]


**Reply by city41** [reply]
Nice. This is exactly the kind of stuff I want to move into. I'm brand new to ghidra and arm though, so I have a lot to learn.

I'm guessing since the `DAT`s have addresses in 0x3000000 they are variables in IRAM.

**Reply by city41** [reply]
@devtronic why did you delete your comment?

**Reply by devtronic** [reply]
I checked the legal situation about this. In Germany we're "allowed to decompile a software to create an interface to another system" but we're "not allowed to share the collected knowledge with third party persons".
